src/Authentication.ts

```diff
@@ -1,231 +1,365 @@
+  }
+
+  /**
+   * Setter method used to add a new two-factor provider.
+   *
+   * @param {TwoFactorProviderInterface} payload
+   */
+  addTwoFactorProvider (payload: TwoFactorProviderInterface): void { // PITANJE da ostavim da prepise vrijednost ako vec postoji ili da radim provjeru
+    this.twoFactorProviders.set(payload.provider, payload)
+  }
+
+  /**
+   * Getter method used to fetch a certain two-factor provider.
+   *
+   * @param {string} provider
+   * @return {TwoFactorProviderInterface}
+   * @throws
+   */
+  getTwoFactorProvider (provider: string): TwoFactorProviderInterface {
+    const result = this.twoFactorProviders.get(provider)
+
+    if (result === undefined) {
+      throw new AuthenticationError('provider not found', { name: 'AuthenticationError', statusCode: 500 })
+      // PITANJE da napravim try catch i throwam kako bi handleError() metoda rijesila error
+    }
+
+    return result
+  }
+
+  /**
+   * Register user.
+   *
+   * @param {RegistrationData} payload
+   * @return {Promise<{ user: AuthenticableUser, twoFactorUser?: AuthenticableTwoFactorUser, qrCode?: string }>}
+   * @throws
+   */
+  async register (payload: RegistrationData): Promise<{ user: AuthenticableUser, twoFactorUser?: AuthenticableTwoFactorUser, qrCode?: string }> {
+    try {
+      this.validationAdapter.registration(payload)
+
+      const user = await this.storageAdapter.register(payload)
+      const result: { user: AuthenticableUser, twoFactorUser?: AuthenticableTwoFactorUser, qrCode?: string } = { user }
+
+      if (payload.twoFactorAuthenticationProvider !== undefined) {
+        const { twoFactorUser, qrCode } = await this.registerTwoFactorUser(user, payload.twoFactorAuthenticationProvider)
+        result.twoFactorUser = twoFactorUser
+        result.qrCode = qrCode
+      }
+
+      return result
+    } catch (error: any) {
+      throw this.errorAdapter.handleError(error)
+    }
+  }
+
+  /**
+   * Login user.
+   *
+   * @param {LoginData} payload
+   * @return {Promise<string>}
+   * @throws
+   */
+  async login (payload: LoginData): Promise<string> {
+    try {
+      this.validationAdapter.login(payload)
+
+      const { user, twoFactorUsers } = await this.storageAdapter.login(payload)
+
+      if (twoFactorUsers.length !== 0) {
+        if (payload.twoFactorAuthenticationData === undefined) {
+          throw new AuthenticationError('invalid credentials', { name: 'AuthenticationError', statusCode: 401 })
+        }
+
+        const twoFactorUser = twoFactorUsers.find(user => user.provider === payload.twoFactorAuthenticationData?.provider)
+
+        if (twoFactorUser === undefined) {
+          throw new AuthenticationError('invalid credentials', { name: 'AuthenticationError', statusCode: 401 })
+        }
+
+        const twoFactorProvider = this.getTwoFactorProvider(payload.twoFactorAuthenticationData.provider)
+        twoFactorProvider.verify(twoFactorUser, payload.twoFactorAuthenticationData.code)
```

Comment:
Could we move this part of the block into a separate method like `verifyTwoFactor` or something like that?

Reply:
Refactored the login method, extracted the 2fa logic to a new method.